From 02cfa453fa0cbd15d108760c1dfc109307ae67f0 Mon Sep 17 00:00:00 2001
From: Franz Flasch <franz.flasch@gmx.at>
Date: Tue, 12 Jan 2021 14:34:58 +0000
Subject: [PATCH 3/6] NPTL fix

---
 package/uclibc/Config.in | 2 +-
 1 file changed, 1 insertion(+), 1 deletion(-)

diff --git a/package/uclibc/Config.in b/package/uclibc/Config.in
index e59fef3..66c2fec 100644
--- a/package/uclibc/Config.in
+++ b/package/uclibc/Config.in
@@ -47,7 +47,7 @@ choice
 
 config BR2_PTHREADS_NATIVE
 	bool "Native POSIX Threading (NPTL)"
-	depends on BR2_USE_MMU
+	depends on BR2_USE_MMU && !BR2_RISCV_64
 	select BR2_TOOLCHAIN_HAS_THREADS
 	select BR2_TOOLCHAIN_HAS_THREADS_NPTL
 
-- 
2.25.1

